src/Mono.Android/Xamarin.Android.Net/AndroidClientHandler.cs

```diff
@@ -272,26 +277,40 @@ string EncodeUrl (Uri url)
 				if(Logger.LogNet)
 					Logger.Log (LogLevel.Info, LOG_APP, " cancelled");
 
-				httpConnection.Disconnect ();
+				await DisconnectAsync (httpConnection).ConfigureAwait (continueOnCapturedContext: false);
 				cancellationToken.ThrowIfCancellationRequested ();
 			}
-			cancellationToken.Register (httpConnection.Disconnect);
 
-			if (httpConnection.DoOutput) {
-				using (var stream = await request.Content.ReadAsStreamAsync ()) {
-					await stream.CopyToAsync(httpConnection.OutputStream, 4096, cancellationToken)
-						.ConfigureAwait(false);
+			CancellationTokenRegistration cancelRegistration = default (CancellationTokenRegistration);
+			HttpStatusCode statusCode = HttpStatusCode.OK;
+			Uri connectionUri = null;
+
+			try {
+				cancelRegistration = cancellationToken.Register (() => {
+					DisconnectAsync (httpConnection).ContinueWith (t => {
+						if (t.Exception != null)
+							Logger.Log (LogLevel.Info, LOG_APP, $"Disconnection exception: {t.Exception}");
+					}, TaskScheduler.Default);
+				}, useSynchronizationContext: false);
+
+				if (httpConnection.DoOutput) {
+					using (var stream = await request.Content.ReadAsStreamAsync ()) {
+						await stream.CopyToAsync(httpConnection.OutputStream, 4096, cancellationToken)
+							.ConfigureAwait(false);
+					}
 				}
+
+				statusCode = await Task.Run (() => (HttpStatusCode)httpConnection.ResponseCode).ConfigureAwait (false);
+				connectionUri = new Uri (httpConnection.URL.ToString ());
+			} finally {
+				cancelRegistration.Dispose ();
 			}
 
 			if (cancellationToken.IsCancellationRequested) {
-				httpConnection.Disconnect();
+				await DisconnectAsync (httpConnection).ConfigureAwait (continueOnCapturedContext: false);
```

Comment:
The correct approach is to always use `.ConfigureAwait (false)` if you want the indicate that you **do not **mind if the subsequent code runs on a threadpool thread, for example https://github.com/xamarin/md-addins/blob/55d85664e677631e647f4aa28597a85193194127/Xamarin.Designer.iOS/MonoTouch.Design.Shared/Commands/BinaryMessageFormatter.cs#L96-L104

There are two reasons why this is important:
1) async calls *may* complete synchronously. If they do then the value of `.ConfigureAwait (false)` is ignored because a continuation is never scheduled. For example:
```
await Task.FromResult (1).ConfigureAwait (false);
Console.WriteLine ("I never left this thread");
```

Or for a more real world example: https://github.com/xamarin/md-addins/blob/55d85664e677631e647f4aa28597a85193194127/Xamarin.Designer.iOS/MonoTouch.Design.Shared/Commands/BinaryMessageFormatter.cs#L65

That call *could* complete synchronously if the socket already has at least 4 bytes in it's buffer, which means i definitely need the `ConfigureAwait (false)` on the second `ReadAsync` call otherwise it will sync back to the UI thread unnecessary.


The second reason is the code you invoke *could* set SyncronizationContext.Current and you probably don't want to accidentally schedule yourself on that. There are valid usecases for setting that value for arbitrary threads.